web/src/message_view.js

```diff
@@ -432,10 +433,15 @@ export function show(raw_terms, opts) {
             if (message_store.get(target_message_id) !== undefined) {
                 id_info.target_id = target_message_id;
             }
-            // If the filter has a `with` operator, we may have a change
-            // in the view from usual filter, hence, we change hash after
-            // updating the view.
-            opts.change_hash = false;
+            // If we have a `with` operator, we may redirect from original narrow.
+            // However, if we are sure that we might end up at initial hash,
+            // we can use `nonredirectional` opts.
+            // This is used to avoid obstruction of using last unread as anchor,
+            // since calling hashchange an extra time calls the narrow.activate
+            // once more.
+            if (!opts.nonredirectional) {
+                opts.change_hash = false;
+            }
```

Comment:
@amanagr The reason I added this block was -- when the topic link was redirectional (old topic link is being used), and you are loading the messages from what is available locally, an additional `change_hash` is called on the updated list, which is affecting the blue box position (because blue box after the view changes first time would be on first unread. When the additional `change_hash` is called, it is calling `show()` again, with messages being marked as read -- blue box on last message).

Reply:
I'm not sure what section of logic needs to be updated, so that this manipulation could be removed safely, without the blue box position changing in above case mentioned.

Reply:
ya, I added thus section as a workaround because of some bug. I guess I found a better way to approach it. I'll update it.

Reply:
updated!